data_steward/analytics/cdr_ops/combined.py

```diff
@@ -0,0 +1,470 @@
+
+,pid_ppi AS
+ (SELECT
+   person_id
+   
+  -- use latest ppi record for most rigid check --
+  ,MAX(o.observation_date) AS max_ppi_date
+  
+  -- in case of multiple death rows, use earliest for most rigid check --
+  ,MIN(d.death_date)       AS min_death_date
+
+  FROM `{DATASET_ID}.death` d
+   JOIN `{DATASET_ID}.observation` o
+    USING (person_id)
+   JOIN ppi_concept c
+    ON o.observation_source_concept_id = c.concept_id
+  GROUP BY person_id)
+
+SELECT * 
+FROM pid_ppi
+WHERE min_death_date < max_ppi_date
+'''
+q(query)
+
+# ## Consent required for EHR Data
+# If EHR records are found for participants who have not consented this may 
+# indicate a bug in the pipeline as these should have been removed. These 
+# records should also be reported to EHR Ops so that sites may cease to send 
+# the information.
+
+# +
+EHR_CONSENT_PERMISSION_CONCEPT_ID = 1586099
+YES_CONCEPT_ID = 1586100
+
+tpl = JINJA_ENV.from_string('''
+-- For participants who do NOT satisfy EHR consent requirements as determined 
+-- by the temp table below, this dynamic query will provide the      --
+-- table_name, person_id, and hpo_id where any EHR records are found --
+DECLARE query DEFAULT
+(SELECT
+
+ -- Generate a subquery for any table containing HPO-submitted data                    --
+ -- which finds any HPO-submitted rows whose pids are not in the consent lookup table. --
+
+ STRING_AGG(
+     '(SELECT DISTINCT '
+      || '"' || table_name || '" AS table_name '
+      || ',t.person_id           AS person_id '
+      || ',m.src_hpo_id          AS hpo_id '
+      || 'FROM `' || table_schema || '.' || table_name || '` t '
+      || 'JOIN `' || table_schema || '.' || table_id || '` m ' 
+      || 'USING (' || table_name ||'_id) '
+      || 'LEFT JOIN consented c '
+      || ' USING (person_id)'
+      || 'WHERE m.src_hpo_id <> "rdr" AND c.person_id IS NULL)'
+   , ' UNION ALL ')
+ FROM `{{DATASET_ID}}.INFORMATION_SCHEMA.COLUMNS` c
+ JOIN `{{DATASET_ID}}.__TABLES__` t
+  ON t.table_id = '_mapping_' || c.table_name
+ WHERE column_name = 'person_id'
+   AND t.row_count > 0
+);
+
+-- PIDs whose last EHR consent response was affirmative --
+CREATE OR REPLACE TEMP TABLE consented AS
+( 
+  WITH 
+
+  -- For each pid, the latest EHR consent record --
+  pid_last_consent_response AS
+  (SELECT DISTINCT 
+     p.person_id
+    ,LAST_VALUE(o.value_source_concept_id)
+      OVER (
+       PARTITION BY o.person_id
+       ORDER BY o.observation_datetime
+       ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
+      ) AS last_value_source_concept_id
+   FROM `{{DATASET_ID}}.person` p
+    JOIN `{{DATASET_ID}}.observation` o
+     ON p.person_id = o.person_id
+   WHERE observation_source_concept_id = {{EHR_CONSENT_PERMISSION_CONCEPT_ID}})
+
+  -- Only store pids where the latest response is YES --
+  SELECT 
+   person_id
+  FROM pid_last_consent_response
+  WHERE 
+      last_value_source_concept_id = {{YES_CONCEPT_ID}}
+);
+
+EXECUTE IMMEDIATE query;
```

Comment:
Per @lrwb-aou , `DECLARE` the variable `query` at the script start, and `SET` it **after** creating the temp table; it's easier to follow that way.

```suggestion
DECLARE query STRING;

-- PIDs whose last EHR consent response was affirmative --
CREATE OR REPLACE TEMP TABLE consented AS
( 
  WITH 

  -- For each pid, the latest EHR consent record --
  pid_last_consent_response AS
  (SELECT DISTINCT 
     p.person_id
    ,LAST_VALUE(o.value_source_concept_id)
      OVER (
       PARTITION BY o.person_id
       ORDER BY o.observation_datetime
       ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
      ) AS last_value_source_concept_id
   FROM `{{DATASET_ID}}.person` p
    JOIN `{{DATASET_ID}}.observation` o
     ON p.person_id = o.person_id
   WHERE observation_source_concept_id = {{EHR_CONSENT_PERMISSION_CONCEPT_ID}})

  -- Only store pids where the latest response is YES --
  SELECT 
   person_id
  FROM pid_last_consent_response
  WHERE 
      last_value_source_concept_id = {{YES_CONCEPT_ID}}
);

SET query = (
SELECT
 -- Generate a subquery for any table containing HPO-submitted data                              --
 -- which finds any HPO-submitted rows whose pids are not in the consent lookup table. --

 STRING_AGG(
     '(SELECT DISTINCT '
      || '"' || table_name || '" AS table_name '
      || ',t.person_id           AS person_id '
      || ',m.src_hpo_id          AS hpo_id '
      || 'FROM `' || table_schema || '.' || table_name || '` t '
      || 'JOIN `' || table_schema || '.' || table_id || '` m ' 
      || 'USING (' || table_name ||'_id) '
      || 'LEFT JOIN consented c '
      || ' USING (person_id)'
      || 'WHERE m.src_hpo_id <> "rdr" AND c.person_id IS NULL)'
   , ' UNION ALL ')
 FROM `{{DATASET_ID}}.INFORMATION_SCHEMA.COLUMNS` c
 JOIN `{{DATASET_ID}}.__TABLES__` t
  ON t.table_id = '_mapping_' || c.table_name
 WHERE column_name = 'person_id'
   AND t.row_count > 0
);

EXECUTE IMMEDIATE query;
```